addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -192,105 +226,174 @@ func _after_tick(_delta, _tick):
-	var sanitized = {}
-	for property in state:
-		var pe = _property_cache.get_entry(property)
-		var value = state[property]
-		var state_owner = pe.node.get_multiplayer_authority()
+	# DEBUG
+	if (root.player_id == 1):
+		_logger.info("Client received tick %s for brawler 1" % received_tick)
+		if (_states.has(received_tick - 1) == false):
+			_logger.warning("non-linear states for brawler 1. Received tick %s while latest tick is %s" % [received_tick, _latest_state_tick])
+
+	# It is guaranteed that the first states are full, until the client sends acks to the server
+	# which enables the server to start sending diffs
+	if (not _has_received_full_state):
+		_has_received_full_state = true
+		_receive_full_state_ack.rpc_id(1, received_tick)
+
+	if (received_state.size() > 0):
+		set_received_state(received_state, received_tick, multiplayer.get_remote_sender_id())
+	elif (NetworkRollback.enable_state_diffs):
+		if (_states.has(received_tick - 1) == false):
+			_logger.warning("non-linear diff states. Received tick %s while latest tick is %s" % [received_tick, _latest_state_tick])
+		if (_states.is_empty()):
+			_logger.error("Received a state without any properties from %s, but missing previous states!" % multiplayer.get_remote_sender_id())
+			breakpoint
+			return
 		
-		if state_owner != sender:
+		for picked_tick in range(_latest_state_tick, received_tick):
+			_states[picked_tick + 1] = _states[picked_tick]
+		_latest_state_tick = received_tick
+
+	else:
+		_logger.error("Received a state without any properties from %s, but diff states which should make it possible, are disabled!" % multiplayer.get_remote_sender_id())
+		breakpoint
+
+func set_received_state(received_state: Dictionary, received_tick: int, sender_id: int):
+	var sanitized: Dictionary = {}
+	var missing_property: bool = false
+	for property_entry in _record_state_props:
+		var state_owner_id: int = property_entry.node.get_multiplayer_authority()
+		if state_owner_id != sender_id:
 			_logger.warning("Received state for node owned by %s from %s, sender has no authority!" \
-				% [state_owner, sender])
+				% [state_owner_id, sender_id])
 			continue
-		
-		sanitized[property] = value
+
+		if (received_state.has(property_entry.to_string()) == false):
+			missing_property = true
+			continue
+
+		var value = received_state[property_entry.to_string()]
+		sanitized[property_entry.to_string()] = value
+
+	## Detect if new property is added on runtime and sent
+	var new_property: bool = false
+	for property in received_state:
+		for property_entry in _record_state_props:
+			if (property_entry.to_string() == property):
+				continue
+		new_property = true
+		break
 	
+	# Duplicates the previous state(s), including the current one
+	# Also fixes packet loss (e.g. a state missing)
+	for picked_tick in range(_latest_state_tick, received_tick):
+		#Check for the very first latest_state_tick which doesn't have a corresponding state
+		if (_states.has(picked_tick)):
```

Comment:
This if check is required because I found a bug with `_latest_state_tick` which exists prior to this PR as well, even with the starting order fix of #284 and I will open an issue for that, but until then this if check saves the day (consider it a hack until that issue is resolved)

To synopsize the issue: `_latest_state_tick` is initialized inside `process_settings()` or `ready`, but wherever, its value is always `NetworkTime.tick - 1`. I changed this to `NetworkTime.tick` as it makes more sense. But even this value is untrue because the latest state is set only if we are the authority (locally at `_record_tick`) or at `_submit_state` if we are not the authority. So it should have the value -1 instead of `NetworkTime.tick` and let this variable tick value be set after its actual state is set.

By setting the value to -1 as it should be, causes some other problems, so yeah I opted for this if check :sweat_smile: